### PR TITLE
Refactor truelearn

### PR DESCRIPTION
**Pull request by yuxqiu**
Guidelines

- Item 10: Prevent Repetition with Assignment Expressions
- Item 12: Avoid Striding and Slicing in a Single Expression
- Item 25: Enforce Clarity with Keyword-Only and Positional-Only Arguments
- Item 27: Use Comprehensions Instead of map and filter
- Item 32: Consider Generator Expressions for Large List Comprehensions
- Item 74: Consider memoryview and bytearray for Zero-Copy Interactions with bytes
- Item 87: Define a Root Exception to Insulate Callers from APIs

Next steps (in two new PRs)

- refactor visualisations + add tests
- add examples of visualisations

### CODE REVIEW

**Comment by yuxqiu** [comment]
@KD-7 ready for review.

**Comment by KD-7** [comment]
> * Define a Root Exception to Insulate Callers from APIs

image
We miss a test case covering this

**Comment by yuxqiu** [comment]
> > * Define a Root Exception to Insulate Callers from APIs
> 
> image We miss a test case covering this

I will fix all the tests in a new PR.

**Review by KD-7 — CHANGES_REQUESTED** [review]

File: truelearn/datasets/_base.py [hunk 1]
```diff
@@ -22,12 +24,13 @@ class RemoteFileMetaData:
...
-            chunk = file.read(chunk_size)
-            if not chunk:
-                break
-            hash_val.update(chunk)
+    buffer = bytearray(chunk_size)
+    write_view = memoryview(buffer)
+
+    with open(filepath, "rb", buffering=0) as file:
+        # 3.8+: use walrus operator instead
+        for bytes_read in iter(lambda: file.readinto(write_view), 0):
```
[the remark this answers is not in the document]
Comment (yuxqiu) [reply truelearn/datasets/_base.py:32]: I suspect this is a bug in Pycharm, as this works fine on my side and passes all the tests.

**Review by KD-7 — APPROVED** [review]

**Comment by KD-7** [comment]
@yuxqiu Delete the branch once, the other PRs target branch has been updated to main

**Comment by yuxqiu** [comment]
Done.

**Pull request merged**